packages/@lwc/engine-core/src/framework/stylesheet.ts

```diff
@@ -209,12 +208,8 @@ export function createStylesheet(vm: VM, stylesheets: string[]): VNode | null {
         for (let i = 0; i < stylesheets.length; i++) {
             insertGlobalStylesheet(stylesheets[i]);
         }
-    } else if (ssr || isHydrating()) {
```

Comment:
This is the concerning part. It seems that during rehydration, the style vnode is very different from a regular rendering, and the diffing algo has special behavior for that... I don't know what the deal is... but I will like to know.

If it turns out that this is the case, maybe then the hydration process will NOT necessary try to preserve the original style, but just go with the replacement of it since that will NOT blink due to the insertion/removal.

Reply:
@caridy The reason is that, in non-SSR, we do two things differently:

1. Use Constructable Stylesheets where supported
2. Hoist style nodes to the nearest native shadow root (e.g. for light DOM styles)

Neither is possible in SSR. (The first because there's no browser standard for it, and the second because we only do one pass through the DOM tree during the rendering.) So we just emit inline `<style>` nodes.

For hydration, we can potentially work around this. When hydrating, we can just remove the inline `<style>`s and do our normal flow. I'm also working on a separate PR to properly remove stylesheets when unused: #2797.

Reply:
FWIW in #2797 I am also getting rid of this usage of `isHydrating`, but for a different reason. So if that PR gets merged first, then you won't have to remove it here.